### PR TITLE
[MM-45990] log JSON parsing errors in API4

### PR DESCRIPTION
**Pull request by noxer**
#### Summary
Errors during the parsing of JSON input were logged or returned very inconsistently. This PR adds a method to log a JSON decode error to the context and uses it whenever JSON input could not be decoded. It also uses the context logger for JSON encoding errors instead of the global `mlog` instance.

#### Ticket Link
[MM-45990](https://mattermost.atlassian.net/browse/MM-45990)
partial [MM-45991](https://mattermost.atlassian.net/browse/MM-45991)

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: web/context.go [hunk 1]
```diff
@@ -221,6 +221,10 @@ func (c *Context) SetInvalidParam(parameter string) {
 	c.Err = NewInvalidParamError(parameter)
 }
 
+func (c *Context) SetInvalidJSONParam(parameter string, err error) {
```
Comment (isacikgoz) [inline web/context.go:224]: This is not specific to json errors actually right? Maybe we can use a more generic name. 0/5

**Review by isacikgoz — APPROVED** [review]
Looks good, just a non-blocking suggestion.

File: web/context.go (hunk 1 above)
Comment (noxer) [reply web/context.go:224]: I like the idea, how would you name the method?

File: model/utils.go [hunk 2]
```diff
@@ -218,10 +217,32 @@ type AppError struct {
...
+		sb.WriteString(er.DetailedError)
+	}
+
+	// render all wrapped errors
+	err := er.wrapped
+	for err != nil {
+		sb.WriteString(", ")
+		sb.WriteString(err.Error())
+		err = errors.Unwrap(err)
+	}
```
Comment (agnivade) [inline model/utils.go:243]: Can we have a test for this? :)

File: api4/integration_action.go [hunk 3]
```diff
@@ -24,7 +25,10 @@ func doPostAction(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 
 	var actionRequest model.DoPostActionRequest
-	json.NewDecoder(r.Body).Decode(&actionRequest)
+	err := json.NewDecoder(r.Body).Decode(&actionRequest)
```
Comment (agnivade) [inline api4/integration_action.go:28]: Nice!

File: api4/saml.go [hunk 4]
```diff
@@ -273,14 +273,24 @@ func resetAuthDataToEmail(c *Context, w http.ResponseWriter, r *http.Request) {
...
-	w.Write(b)
+
+	n := struct {
+		NumAffected int `json:"num_affected"`
+	}{
+		NumAffected: numAffected,
+	}
+
+	b, _ := json.Marshal(n)
+	_, err := w.Write(b)
```
Comment (agnivade) [inline api4/saml.go:292]: Usually we do, `json.NewEncoder(w).Encode(...)` to do it in a single step. 0/5 as long as we log the error.

File: shared/mail/inbucket.go [hunk 5]
```diff
@@ -74,11 +74,8 @@ func GetMailBox(email string) (results JSONMessageHeaderInbucket, err error) {
 
 	var record JSONMessageHeaderInbucket
 	err = json.NewDecoder(resp.Body).Decode(&record)
-	switch {
-	case err == io.EOF:
```
Comment (agnivade) [inline shared/mail/inbucket.go:78]: lol this was strange

**Review by agnivade — COMMENTED** [review]
Lots of good stuff!

File: api4/saml.go (hunk 4 above)
Comment (noxer) [reply api4/saml.go:292]: Right, will change that.

File: model/utils.go (hunk 2 above)
Comment (noxer) [reply model/utils.go:243]: Added a test for the error string rendering.

File: api4/integration_action.go (hunk 3 above)
Comment (noxer) [reply api4/integration_action.go:28]: There were a few of those.

File: api4/saml.go (hunk 4 above)
Comment (noxer) [reply api4/saml.go]: Fixed

File: shared/mail/inbucket.go (hunk 5 above)
Comment (noxer) [reply shared/mail/inbucket.go:78]: I thought it was maybe some left over code from when it was handled differently?

Comment (agnivade) [reply shared/mail/inbucket.go:78]: yeah possibly

**Review by agnivade — APPROVED** [review]
Nice!

**Pull request merged**